src/lnp/session/noise/conduit.rs

```diff
@@ -0,0 +1,630 @@
+use super::{chacha, hkdf};
+use crate::lnp::session::transcoders::{Decrypt, Encrypt, Transcode};
+use amplify::Bipolar;
+
+use super::handshake::HandshakeError;
+use chacha20poly1305::aead::Error;
+use std::borrow::Borrow;
+
+pub type SymmetricKey = [u8; 32];
+
+/// Maximum Lightning message data length according to
+/// [BOLT-8](https://github.com/lightningnetwork/lightning-rfc/blob/v1.0/08-transport.md#lightning-message-specification)
+/// and [BOLT-1](https://github.com/lightningnetwork/lightning-rfc/blob/master/01-messaging.md#lightning-message-format):
+const LN_MAX_MSG_LEN: usize = ::std::u16::MAX as usize; // Must be equal to 65535
+
+const MESSAGE_LENGTH_HEADER_SIZE: usize = 2;
+const TAGGED_MESSAGE_LENGTH_HEADER_SIZE: usize =
+    MESSAGE_LENGTH_HEADER_SIZE + chacha::TAG_SIZE;
+
+const KEY_ROTATION_INDEX: u32 = 1000;
+
+pub struct Encryptor {
+    sending_key: SymmetricKey,
+    sending_chaining_key: SymmetricKey,
+    sending_nonce: u32,
+}
+
+impl Encryptor {
+    pub fn encrypt(&mut self, buffer: &[u8]) -> Result<Vec<u8>, Error> {
+        if buffer.len() > LN_MAX_MSG_LEN {
+            panic!("Attempted to encrypt message longer than 65535 bytes!");
+        }
+
+        let length = buffer.len() as u16;
+        let length_bytes = length.to_be_bytes();
+
+        let mut ciphertext = vec![
+            0u8;
+            TAGGED_MESSAGE_LENGTH_HEADER_SIZE
+                + length as usize
+                + chacha::TAG_SIZE
+        ];
+
+        chacha::encrypt(
+            &self.sending_key,
+            self.sending_nonce as u64,
+            &[0; 0],
+            &length_bytes,
+            &mut ciphertext[..TAGGED_MESSAGE_LENGTH_HEADER_SIZE],
+        )?;
+        self.increment_nonce();
+
+        &chacha::encrypt(
+            &self.sending_key,
+            self.sending_nonce as u64,
+            &[0; 0],
+            buffer,
+            &mut ciphertext[TAGGED_MESSAGE_LENGTH_HEADER_SIZE..],
+        )?;
+        self.increment_nonce();
+
+        Ok(ciphertext)
+    }
+
+    fn increment_nonce(&mut self) {
+        Conduit::increment_nonce(
+            &mut self.sending_nonce,
+            &mut self.sending_chaining_key,
+            &mut self.sending_key,
+        );
+    }
+}
+
+impl Encrypt for Encryptor {
+    fn encrypt(&mut self, buffer: impl Borrow<[u8]>) -> Vec<u8> {
+        match self.encrypt(buffer.borrow()) {
+            Ok(values) => return values,
+            Err(_) => return Vec::new(),
+        }
+    }
```

Comment:
Later I will refactor `Encrypt` trait such that it may return error. But for now this will work

Reply:
Ya noted the same. I thought of doing the refactoring myself to return `Result` but I thought it would need modification elsewhere so would be better to do it in separate PR.